benches/benches/bevy_ecs/scheduling/schedule.rs

```diff
@@ -57,75 +57,426 @@ pub fn schedule(c: &mut Criterion) {
     group.finish();
 }
 
+/// performs takes a value out of a reference, applies a fn, and puts it back in.
+/// stores a temporary dummy value while performing the operation.
+fn map_with_temp<T>(ptr: &mut T, temp: T, f: impl FnOnce(T) -> T) {
+    let val = std::mem::replace(ptr, temp);
+    *ptr = f(val);
+}
+
 pub fn build_schedule(criterion: &mut Criterion) {
-    // empty system
-    fn empty_system() {}
-
-    // Use multiple different kinds of label to ensure that dynamic dispatch
-    // doesn't somehow get optimized away.
-    #[derive(Debug, Clone, Copy)]
-    struct NumLabel(usize);
-    #[derive(Debug, Clone, Copy, SystemLabel)]
-    struct DummyLabel;
-
-    impl SystemLabel for NumLabel {
-        fn as_str(&self) -> &'static str {
-            let s = self.0.to_string();
-            Box::leak(s.into_boxed_str())
-        }
+    use bevy_ecs::{
+        prelude::*,
+        schedule::{ParallelSystemDescriptor, SystemLabelId},
+    };
+
+    // Simulates a plugin that has a decent number of systems.
+    // Systems have interdependencies within plugins,
+    // as well as with public labels exported by other plugins.
+    // Also, sometimes entire plugins have dependencies with one another, via the plugin's own label.
+    struct Plugin {
+        label: SystemLabelId,
+        systems: [ParallelSystemDescriptor; 20],
+        pub_labels: [SystemLabelId; 4],
     }
 
-    let mut group = criterion.benchmark_group("build_schedule");
-    group.warm_up_time(std::time::Duration::from_millis(500));
-    group.measurement_time(std::time::Duration::from_secs(15));
+    #[derive(SystemLabel)]
+    struct PluginLabel<const I: usize>;
+
+    #[derive(SystemLabel)]
+    enum PubLabel<const P: usize> {
+        Short,
+        LongName,
+        ReallyLongName,
+        ReallyReallySuperLongName,
+    }
+
+    fn my_system<const P: usize, const I: usize>() {}
+
```

Comment:
I would like the number of plugins and systems to be pulled out into constants here.

Reply:
The problem with this is that, afaik, every single const generic type needs to be explicitly named somewhere. You can't loop over const generics. I can make it prettier with macros, but I don't think it's possible to extract everything into constants.

Reply:
Hmm, that's surprising 🤔 I'm not familiar enough with the details of macros to know if what I'm asking for is possible. We should see if we can get more expert opinions.

Reply:
Initial digging is promising:
1. https://stackoverflow.com/questions/67038789/writing-const-generic-enum-combinations-using-macro
2. https://stackoverflow.com/a/70233895

And bingo: https://docs.rs/seq-macro/latest/seq_macro/